### ISSUE TITLE
Specialization constant expressions and ternary operator

### ISSUE THREAD

**Issue by tksuoran** [KhronosGroup/Vulkan-Docs#424]
Let's have layout (constant_id = 0) const int POINT_LIGHT_COUNT = 0. Since zero sized arrays are not valid, I thought I could work around this by using POINT_LIGHT_COUNT > 0 ? POINT_LIGHT_COUNT : 1 as array size for a varying. Unfortunately ternary operator is not allowed operator for specialization constant expressions. This feels strange, because I can still use int(POINT_LIGHT_COUNT > 0) * POINT_LIGHT_COUNT + int(POINT_LIGHT_COUNT <= 0) * 1 as array size.

Could we allow also ternary operator with specialization constant expressions?

Could we allow zero sized arrays, if they are not referenced at all?

Note that this is not only about the default value being zero, which as arrays size is rejected by glslangvalidator; I want also runtime specialization constant value 0 to work.

Initially created at https://github.com/KhronosGroup/glslang/issues/649

**Comment by NicolBolas** [comment]
Wait: if SPIR-V allows `OpSelect` in specialization constants, how exactly is this the Vulkan specification's fault? If you're getting an error from glslang, then they're the one's who need to fix their compiler.

As for the stuff about zero-sized arrays... why should they be allowed? It sounds like extra work for the IHV's compilers.

**Comment by tksuoran** [comment]
Ternary operator is disallowed in https://www.khronos.org/registry/vulkan/specs/misc/GL_KHR_vulkan_glsl.txt which is Khronos specification that glslang has to respect.

Let's say I pass vector from surface point to point light position in varying arrays. As long as zero sized arrays are not allowed, I have to use COUNT > 0 : COUNT : 1 expression if I want to use specialization constants alone to specialize different number of point lights, and I will waste one varying doing this. It is possible to use other means (preprocessor defines etc.) to optimize the case of COUNT = 0 without wasting a varying.

If zero sized arrays were allowed, it will make specialization constants slightly more useful and powerful feature. I would have imagined that it is no extra work for IHV's to simply drop rejecting zero sized arrays, but of course I could be wrong.

Just to clarify, does Vulkan / SPIR-V / Vulkan GLSL specification specify that Vulkan runtime should reject shader when specialization expression would result zero sized array?

**Comment by NicolBolas** [comment]
> Ternary operator is disallowed in https://www.khronos.org/registry/vulkan/specs/misc/GL_KHR_vulkan_glsl.txt which is Khronos specification that glslang has to respect.

Ah, so it's just a bug in KHR_vulkan_glsl, not properly exposing all of what SPIR-V allows.

> I would have imagined that it is no extra work for IHV's to simply drop rejecting zero sized arrays

You assume that the compiler could otherwise function reasonably with a zero-sized array. The very concept requires answering a lot of questions.

It should be noted that _even C and C++_ don't allow zero-sized arrays (though admittedly in C, this is a common compiler extension).

> Just to clarify, does Vulkan / SPIR-V / Vulkan GLSL specification specify that Vulkan runtime should reject shader when specialization expression would result zero sized array?

The Vulkan runtime only accepts legal SPIR-V. And the SPIR-V command [`OpTypeArray` requires the length to be one](https://www.khronos.org/registry/spir-v/specs/1.1/SPIRV.html#OpTypeArray). GLSL itself has always disallowed zero-sized arrays.

**Comment by krOoze** [comment]
One could make a case, that some of the simple functions like `max` should be allowed too.

**Comment by johnkslang** [comment]
This is now being discussed internally in Khronos, as three issues:

1. have GLSL support the `?:` operator, which is localized to the specification and front-end compilers
2. support 0-sized arrays, which impacts shipping drivers, and hence can only be tied to some future release of Vulkan
3. white-list built-in functions for specialization constants, which also impacts shipping drivers

Note that once `?:` is added, it can be used for `max`.

**Comment by johnkslang** [comment]
Khronos has resolved this as follows:  It has decided to support the `?:` operator.

Glslang now supports this, currently on the emit-opSelect side branch, but soon to be master.

Note this nicely enough avoids need for things like max() to handle specialization constants:

```
int ternayArray1[a > b ? a : b]; // for a and b specialization constants
```

0-sized arrays are troublesome enough for enough compiler code, in layers everywhere, that they will not be supported anytime soon.

**Comment by krOoze** [comment]
> Note this nicely enough avoids need for things like max() to handle specialization constants

If it is so trivial, isn't it a reason to allow things like `max()` there?

**Comment by NicolBolas** [comment]
@krOoze But it's not trivial in the case of `max`, because that function is actually an OpenGL extended SPIR-V opcode. It is currently not listed in the list of opcodes that can manipulate specialization constants. Which means it would have to be *added* to that list.

See, `OpSelect` [was already on that list](https://www.khronos.org/registry/spir-v/specs/1.1/SPIRV.html#OpSpecConstantOp), so glslang simply needed to be extended to follow the SPIR-V spec more correctly. What you're talking about would require _changing_ the actual SPIR-V specification, along with updates to implementations.

Now yes, glslang _could_ start converting all uses of `max()` to `OpSelect`. However, that would perhaps be surprising to users; calls to GLSL library functions should make the appropriate calls to the SPIR-V extended opcodes.

**Comment by krOoze** [comment]
> calls to GLSL library functions should make the appropriate calls to the SPIR-V extended opcodes.

Why? Similarity between GLSL and SPIR-V is more of a accident of birth than deliberate 1:1 mapping.
How do you propose HLSL should be mapped? How should some possible C++ derivative shader lang be mapped?

**Comment by johnkslang** [comment]
@NicolBolas agreed.

Additionally, even if glslang converted `max()` to `OpSelect`, the source-code's `max()` would still require discussing propagation of the specialization-constant typeness:  Is `5 + max(...)` also a specialization constant? That requires changing other things too, including having this `max()` behavior dependent on the type of operands and agreeing on a specific white-list of GLSL built-ins, and what impact that has on multiple specs., etc.

**Comment by krOoze** [comment]
@johnkslang 
As for propagation, `GL_KHR_vulkan_glsl`:
> An expression formed with specialization constants also behaves in the
shader like a specialization constant, not a like a constant.

Seem'd as easy as adding a word or two to the spec (same as the ternary op). But if it is hard, it's hard. All in due time...

**Comment by johnkslang** [comment]
@krOoze There is a white list of allowed operations that go into making that statement fully true, in the normative language (not in the overview). The change we made for this issue is to add `?:` to that normative white list.

**Comment by johnkslang** [comment]
Re-summarizing: The fixes for `?:` are all done and and either already available or on the way (next spec drop). This is a nice workaround for things like `max()`. If this is truly inadequate, open a new issue, so this one can be closed along with the completion of delivery of the decisions already made for `?:`. Thanks.

**Comment by oddhack** [comment]
Should be fixed in the 1.0.40 spec update.
